skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+
+        fve = np.cumsum(eigenvalues) / np.sum(eigenvalues)
+        if isinstance(n_components, int):
+            if n_components > len(eigenvalues):
+                error_msg = (
+                    "The sample size must be bigger than the number of "
+                    "components"
+                )
+                raise AttributeError(error_msg)
+            n_selected_components = n_components
+        else:
+            # Find the optimal number of components
+            n_selected_components = np.where(fve >= n_components)[0][0] + 1
+
+        eigenvalues = eigenvalues[:n_selected_components]
+        eigenvectors = eigenvectors[:, :n_selected_components]
+
+        lambda_ = h * eigenvalues
+        eigenvectors /= np.sqrt(h)
+
+        eigenvectors = self._normalize_and_orient(
+            eigenvectors, self.t_covariance_.squeeze()
+        )
+
+        phi = self._interpolate_and_normalize_basis(
+            t_eigen, eigenvectors, self.mean_.grid_points[0],
+        )
+
+        return n_selected_components, fve, lambda_, phi.T
+
+    def _get_sigma2(
+        self,
+        h: float,
+        t_eval: NDArrayFloat,
+        cov_coords: NDArrayFloat,
+        cov_values: NDArrayFloat,
+        t_diag: NDArrayFloat,
+        cov_diag: NDArrayFloat,
+        win: NDArrayFloat,
+        domain_range: NDArrayFloat,
+    ) -> float:
+        """
+        Estimate the variance of the covariance matrix.
+
+        Args:
+            h: Bandwidth for the kernel. It is the same one used to smooth the
+                covariance.
+            t_eval: Query points where diagonal is evaluated, expected to be
+                (num eval points x 1)-dimensional.
+            cov_coords: Coordinates of the covariance.
+            cov_values: Values of the covariance.
+            t_diag: Coordinates of the raw covariance diagonal.
+            cov_diag: Values of the raw covariance diagonal.
+            win: Weights for the covariance.
+            domain_range: Domain range of the data.
+
+        Returns:
+            The estimated variance.
+        """
+        smooth_diag = self._mean_lls(
+            h,
+            t_eval,
+            t_diag,
+            cov_diag,
+            self.kernel_cov,
+        )
+
+        rotated_cov_diag = self._rotated_cov_lls(
+            h,
+            t_eval,
+            t_eval,
+            cov_coords,
+            cov_values,
+            win,
+        )
+
+        min_domain, max_domain = domain_range[0]
+        domain_range = max_domain - min_domain
+        a = min_domain + domain_range * self.variance_error_interval[0]
+        b = max_domain - domain_range * (1 - self.variance_error_interval[1])
+
+        x = t_eval.squeeze()
+        y = (smooth_diag - rotated_cov_diag).squeeze()
+
+        # Mask for integration interval [a, b]
+        mask = (x > a) & (x < b)
+
+        # Perform Simpson integration and scale
+        sigma2 = trapezoid(y[mask], x[mask]) * 2 / domain_range
```

Comment:
I would rather build a `FDataGrid` and call its `integrate` method. This would be slower for sure, but this way we use consistently the same quadrature method everywhere. Also it makes sense to use our types when we are dealing with functions, instead of bare arrays.